### DISCUSSION TITLE
导出inference模型只有两个文件，无法使用

### DISCUSSION THREAD

**Discussion by ysmzgddxhhlkb** [PaddlePaddle/PaddleOCR#14279] [Q&A] [answered]
我的转换代码如下：
`# 转换检测模型为推理模型
!python /home/aistudio/PaddleOCR-release-2.8/tools/export_model.py -c "/home/aistudio/PaddleOCR-release-2.8/configs/det/ch_PP-OCRv4/ch_PP-OCRv4_det_teacher.yml" -o Global.pretrained_model="/home/aistudio/output/ch_PP-OCRv4_det_server/best_accuracy.pdparams" Global.save_inference_dir="/home/aistudio/Myinfer_model/new_det/"
# 转换识别模型为推理模型
!python /home/aistudio/PaddleOCR-release-2.8/tools/export_model.py -c "/home/aistudio/PaddleOCR-release-2.8/configs/rec/PP-OCRv4/ch_PP-OCRv4_rec_hgnet.yml" -o Global.pretrained_model="/home/aistudio/output/rec_ppocr_v4_hgnet/best_accuracy.pdparams" Global.save_inference_dir="/home/aistudio/Myinfer_model/new_rec/"`
![09E72268AF6B65246D114BE971F30390](https://github.com/user-attachments/assets/ed224870-7c76-4df9-8689-2c66178bb1dd)

报错：
pre_model\new_det\inference.pdmodel does not exists.

**Comment by GreatV** [comment]
根据您提供的信息以及相关的 GitHub 讨论和问题描述，导出 inference 模型时缺失 `inference.pdmodel` 或其他推理所需文件可能与以下原因相关。以下是可能的原因和解决方案：

---

### **问题分析：**

1. **导出模型步骤问题：**
   - 确保您在使用 `export_model.py` 时，路径和参数正确。特别是 `Global.pretrained_model` 的路径应该指向训练得到的 `best_accuracy.pdparams` 文件，而不是其他文件夹或错误路径。
   - `Global.save_inference_dir` 的路径应为导出推理模型的目标目录，且确保该路径存在并且有写入权限。

2. **PaddlePaddle 和 PaddleOCR 版本不匹配：**
   - 导出模型的工具 `export_model.py` 和 PaddleOCR 的版本是否和您当前安装的 PaddlePaddle 框架版本兼容？从讨论中可以看出，部分用户在 PaddleOCR 2.9 或更高版本中遇到了类似问题。

3. **模型配置文件问题：**
   - 确保您使用的配置文件（如 `ch_PP-OCRv4_det_teacher.yml` 和 `ch_PP-OCRv4_rec_hgnet.yml`）没有误改。例如，某些模型可能需要特定的后处理配置参数。如果配置文件中缺少相关参数，可能导致导出模型失败或文件缺失。

4. **未正确加载预训练模型参数：**
   - 如果 `best_accuracy.pdparams` 文件损坏或与当前配置文件不匹配，可能会导致导出失败。确保训练模型在 `Global.pretrained_model` 指定路径下是完整、正确的。

5. **推理模型文件缺失：**
   - 导出推理模型时，通常会生成以下三个文件：
     - `inference.pdmodel`
     - `inference.pdiparams`
     - `inference.pdiparams.info`
   - 如果缺少上述文件中的任意一个，通常是导出模型时出现问题，比如路径错误、配置问题或者版本不兼容。

---

### **解决方案：**

#### **1. 检查路径和文件完整性**
- 确保以下路径和文件存在：
  ```bash
  /home/aistudio/output/ch_PP-OCRv4_det_server/best_accuracy.pdparams
  /home/aistudio/output/rec_ppocr_v4_hgnet/best_accuracy.pdparams
  ```
- 同时，确认导出的推理模型目录是否正确：
  ```bash
  /home/aistudio/Myinfer_model/new_det/
  /home/aistudio/Myinfer_model/new_rec/
  ```

#### **2. 更新 PaddleOCR 和 PaddlePaddle**
- 如果您使用的是 PaddleOCR 2.8 或更早版本，建议升级到最新版本（例如 2.9 或更高），同时确保 PaddlePaddle 框架版本与之兼容。
  ```bash
  pip install --upgrade paddlepaddle
  pip install --upgrade paddleocr
  ```

#### **3. 确保正确加载训练权重**
- 检查训练过程中生成的权重文件是否完整且能正常加载。如果怀疑文件损坏，可以尝试直接使用 `infer_rec.py` 测试训练模型，以验证其有效性。

#### **4. 修改导出模型的代码**
尝试在导出模型时显式指定 `inference.pdmodel` 和其他文件路径，确保模型正确保存。以下是一个示例命令：
```bash
python tools/export_model.py \
  -c /home/aistudio/PaddleOCR-release-2.8/configs/det/ch_PP-OCRv4/ch_PP-OCRv4_det_teacher.yml \
  -o Global.pretrained_model="/home/aistudio/output/ch_PP-OCRv4_det_server/best_accuracy.pdparams" \
     Global.save_inference_dir="/home/aistudio/Myinfer_model/new_det/"
```

#### **5. 检查日志输出**
- 导出模型时，注意终端输出日志，确认是否有警告或错误提示。例如，如果日志中提示 `load pretrain successful`，说明模型权重加载成功。

#### **6. 替代解决方案**
如果上述方法仍然无法解决问题，可以尝试以下步骤：
- 重新下载官方训练好的权重文件，并尝试直接将其导出为推理模型。
  - 检测模型：[ch_PP-OCRv4_det_server_train.tar](https://paddleocr.bj.bcebos.com/PP-OCRv4/chinese/ch_PP-OCRv4_det_server_train.tar)
  - 识别模型：[en_PP-OCRv4_rec_train.tar](https://paddleocr.bj.bcebos.com/PP-OCRv4/english/en_PP-OCRv4_rec_train.tar)
- 如果导出成功，说明您的训练模型可能存在问题。

---

### **其他用户问题相关性**
根据以下 GitHub 讨论和问题，与您的情况类似的问题已经被提及：
- [Issue 14223](https://github.com/PaddlePaddle/PaddleOCR/issues/14223): 模型转换后缺失 pdmodel。
- [Issue 14242](https://github.com/PaddlePaddle/PaddleOCR/issues/14242): 训练模型 export 后缺少推理文件。
- [Discussion 14088](https://github.com/PaddlePaddle/PaddleOCR/discussions/14088): 导出模型后推理结果不对。

这些讨论表明，缺失 `inference.pdmodel` 和其他文件可能与训练模型本身或路径配置有关。

---

### **总结**
请按照以下步骤逐一排查：
1. 确认路径和文件是否正确。
2. 更新 PaddleOCR 和 PaddlePaddle 到最新版本。
3. 检查训练模型能否正常加载。
4. 使用官方权重文件验证导出是否成功。
5. 检查配置文件的正确性，确保没有误改。

如果问题仍然无法解决，可以将完整的报错日志和导出命令提交到 [PaddleOCR Issues](https://github.com/PaddlePaddle/PaddleOCR/issues) 页面，获取进一步的帮助。

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

**Comment by GreatV** [comment]
可以试试 `export FLAGS_enable_pir_api=0`

**Reply by Rogerlv51** [reply]
> 可以试试 `export FLAGS_enable_pir_api=0`

请问windows系统下怎么办呢

**Comment by ysmzgddxhhlkb** [comment]
感谢您的耐心回复，现在问题已经解决。
原因如下：
是因为用的AiStudio平台没有安装ccache所以导致导出推理模型的时候只有两个文件。
解决：
安装这个ccache，导出就没问题了。






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PaddlePaddle/PaddleOCR"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年11月30日(星期六) 中午11:18
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [PaddlePaddle/PaddleOCR] 导出inference模型只有两个文件，无法使用 (Discussion #14279)





 
可以试试 export FLAGS_enable_pir_api=0

**Reply by GreatV** [reply]
跟 ccache没关系

**Reply by ysmzgddxhhlkb** [reply]
但是同样的配置文件与模型，下载至个人电脑上可以正常导出。以下Aistudio训练警告信息：
![QQ_1732954177375](https://github.com/user-attachments/assets/bc0778b2-8f6b-4b51-9c63-2716e7cc4bcb)

**Reply by GreatV** [reply]
哥，你不同环境paddle框架版本不一样

**Reply by ysmzgddxhhlkb** [reply, marked as answer]
好的吧，我检查了一下Aistudio上面的Paddle版本确实与我本机的Paddle版本不同，害。感谢您的答疑。

**Reply by LiuJiPeng20** [reply]
> 哥，你不同环境paddle框架版本不一样

你好，我的paddle框架版本是2.10，【文本方向分类模型】在执行训练好后的推理模型导出时，推理模型文件夹下只有inference.pdiparams.info、inference.pdmodel、inference.yml，3个文件，缺少inference.pdiparams。请问这是为什么？

代码复现：
PaddleOCR-main>python tools/export_model.py -c configs/cls/cls_mv3.yml -o Global.pretrained_model=./output/cls/mv3/latest Global.save_inference_dir=./models/cls_train/
[2025/04/16 10:07:11] ppocr WARNING: Skipping import of the encryption module.
[2025/04/16 10:07:13] ppocr INFO: load pretrain successful from ./output/cls/mv3/latest
[2025/04/16 10:07:13] ppocr INFO: Export inference config file to ./models/cls_train/inference.yml
Skipping import of the encryption module
I0416 10:07:18.140360  4644 program_interpreter.cc:243] New Executor is Running.
[2025/04/16 10:07:18] ppocr INFO: inference model is saved to ./models/cls_train/inference

**Reply by Rogerlv51** [reply]
> > 哥，你不同环境paddle框架版本不一样
> 
> 你好，我的paddle框架版本是2.10，【文本方向分类模型】在执行训练好后的推理模型导出时，推理模型文件夹下只有inference.pdiparams.info、inference.pdmodel、inference.yml，3个文件，缺少inference.pdiparams。请问这是为什么？
> 
> 代码复现： PaddleOCR-main>python tools/export_model.py -c configs/cls/cls_mv3.yml -o Global.pretrained_model=./output/cls/mv3/latest Global.save_inference_dir=./models/cls_train/ [2025/04/16 10:07:11] ppocr WARNING: Skipping import of the encryption module. [2025/04/16 10:07:13] ppocr INFO: load pretrain successful from ./output/cls/mv3/latest [2025/04/16 10:07:13] ppocr INFO: Export inference config file to ./models/cls_train/inference.yml Skipping import of the encryption module I0416 10:07:18.140360 4644 program_interpreter.cc:243] New Executor is Running. [2025/04/16 10:07:18] ppocr INFO: inference model is saved to ./models/cls_train/inference

我也是一样，windows系统，你解决了吗

**Comment by ysmzgddxhhlkb** [comment]
这边又发现了一个问题，崩溃破防了，哥。求帮忙。
我先是用官方默认的检测和识别模型，发现识别不了δ，所以自己训练了模型识别δ。
然后现在出现了一个很诡异的现象，就是我用官方的模型检测同样的图片，竟然能识别到δ了？？？（此时我没用我训练的模型）
![image](https://github.com/user-attachments/assets/fb6a87fd-3717-4417-b7a9-fef4749c67ff)
![QQ_1732984540225](https://github.com/user-attachments/assets/9e2ffdec-96bb-41b2-9a9f-54942a6d3fed)
是不是有着什么模型缓存之类的东西？导致它使用了我训练的模型？

**Reply by GreatV** [reply]
rapidocr的机制不太了解，paddleocr是没有缓存的，可能你最开始检测和现在检测用的不是同一张图。

**Comment by BeckShi0923** [comment]
<img width="1140" height="198" alt="mismatch" src="https://github.com/user-attachments/assets/d53cf46c-5e4b-4e79-ba1e-fd4f30944ff9" />
用PaddleOCR-release-3.0 训练 ch_PP-OCRv3_rec_slim_train 用premodel 报错，不匹配 什么原因啊
